homeassistant/components/hassio/backup.py

```diff
@@ -0,0 +1,170 @@
+            )
+            for backup in await get_supervisor_client(self._hass).backups.list()
+        ]
+
+    async def async_get_backup(
+        self,
+        backup_id: str,
+        **kwargs: Any,
+    ) -> BaseBackup | None:
+        """Return a backup."""
+        backups = await self.async_list_backups()
+        for backup in backups:
+            if backup.backup_id == backup_id:
+                return backup
+        return None
+
+    def get_backup_path(self, backup_id: str) -> Path:
+        """Return the local path to a backup."""
+        return self._backup_dir / f"{backup_id}.tar"
+
+    async def async_delete_backup(self, backup_id: str, **kwargs: Any) -> None:
+        """Remove a backup."""
+        raise NotImplementedError
+
+
+class SupervisorBackupReaderWriter(BackupReaderWriter):
+    """Class for reading and writing backups in supervised installations."""
+
+    temp_backup_dir = Path("/cloud_backups")
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the backup reader/writer."""
+        self._hass = hass
+
+    async def async_create_backup(
+        self,
+        *,
+        addons_included: list[str] | None,
+        agent_ids: list[str],
+        database_included: bool,
+        backup_name: str,
+        folders_included: list[str] | None,
+        on_progress: Callable[[BackupProgress], None] | None,
+        password: str | None,
+    ) -> tuple[NewBackup, asyncio.Task[tuple[BaseBackup, Path]]]:
+        """Create a backup."""
+        addons_included_set = set(addons_included) if addons_included else None
+        folders_included_set = set(folders_included) if folders_included else None
+
+        client = get_supervisor_client(self._hass)
+        backup = await client.backups.partial_backup(
+            supervisor_backups.PartialBackupOptions(
+                addons=addons_included_set,
+                folders=folders_included_set,  # type: ignore[arg-type]
+                homeassistant=True,
+                name=backup_name,
+                password=password,
+                compressed=True,
+                location=None,
+                homeassistant_exclude_database=not database_included,
+                background=True,
+            )
+        )
+        backup_task = self._hass.async_create_task(
+            self._async_wait_for_backup(backup),
+            name="backup_manager_create_backup",
+            eager_start=False,  # To ensure the task is not started before we return
+        )
+
+        return (NewBackup(backup_id=backup.job_id), backup_task)
+
+    async def _async_wait_for_backup(
+        self, backup: supervisor_backups.NewBackup
+    ) -> tuple[BaseBackup, Path]:
+        """Wait for a backup to complete."""
+        raise NotImplementedError
+
+    async def async_restore_backup(
+        self,
+        backup_id: str,
+        *,
+        agent_id: str,
+        password: str | None,
+        **kwargs: Any,
+    ) -> None:
```

Comment:
The core API is incomplete, it needs to be extended with partial restore options